_doc/man/man3/species.3
.TH "species" 3 "Fri Mar 28 2014" "Version 4.8 (20140327.66)" "CaRNeSS" \" -*- nroff -*-
.ad l
.nh
.SH NAME
species \- 
.PP
This class contains declarations of the species class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <species\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBspecies\fP ()"
.br
.RI "\fI< New species constructor (IN AMOUNT) \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_longInt\fP tmpAmount, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpComplexDegEnh, \fBacs_int\fP tmpComplexCuttingPoint, \fBacs_int\fP tmpEvalueted, \fBacs_double\fP tmpVolume, \fBacs_double\fP tmpK_phospho, \fBacs_int\fP tmpEnergizable, \fBacs_double\fP tmpInflux_rate, \fBacs_int\fP tmpMaxLOut)"
.br
.RI "\fINew species constructor (IN CONCENTRATION) \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_double\fP tmpConcentration, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpComplexDegEnh, \fBacs_int\fP tmpComplexCuttingPoint, \fBacs_int\fP tmpEvalueted, \fBacs_double\fP tmpVolume, \fBacs_double\fP tmpK_phospho, \fBacs_int\fP tmpEnergizable, \fBacs_double\fP tmpInflux_rate, \fBacs_int\fP tmpMaxLOut)"
.br
.RI "\fINew species constructor in case of species structure file upload (IN AMOUNT) \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_longInt\fP tmpAmount, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpComplexDegEnh, \fBacs_int\fP tmpComplexCuttingPoint, \fBacs_int\fP tmpEvalueted, \fBacs_double\fP tmpAge, \fBacs_int\fP tmpReborns, \fBacs_double\fP tmpVolume, \fBacs_longInt\fP tmpNotUsedCatID, \fBacs_longInt\fP tmpNotUsedSubID, \fBacs_double\fP tmpK_phospho, \fBacs_int\fP tmpEnergizable, \fBacs_double\fP tmpInflux_rate, \fBacs_int\fP tmpMaxLOut)"
.br
.RI "\fINew species constructor in case of species structure file upload (IN CONCENTRATION) \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_double\fP tmpConcentration, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpComplexDegEnh, \fBacs_int\fP tmpComplexCuttingPoint, \fBacs_int\fP tmpEvalueted, \fBacs_double\fP tmpAge, \fBacs_int\fP tmpReborns, \fBacs_double\fP tmpVolume, \fBacs_longInt\fP tmpNotUsedCatID, \fBacs_longInt\fP tmpNotUsedSubID, \fBacs_double\fP tmpK_phospho, \fBacs_double\fP tmpKLoadConc, \fBacs_int\fP tmpEnergizable, \fBacs_double\fP tmpInflux_rate, \fBacs_int\fP tmpMaxLOut)"
.br
.RI "\fINew random species constructor\&. \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_longInt\fP tmpAmount, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpComplexProb, \fBacs_double\fP tmpMaxComplexDegKinetic, \fBMTRand\fP &tmp_RandomGenerator, \fBacs_double\fP tmpVolume, \fBacs_double\fP tmpK_phospho, \fBacs_int\fP tmpEnergizable)"
.br
.RI "\fInew Complex species constructor \fP"
.ti -1c
.RI "\fBspecies\fP (\fBacs_longInt\fP tmpID, string tmpSequence, \fBacs_double\fP tmpDiffusionEnh, \fBacs_int\fP tmpSoluble, \fBacs_double\fP tmpMaxComplexDegKinetic, \fBacs_int\fP tmpCuttingPoint, \fBMTRand\fP &tmp_RandomGenerator, \fBacs_longInt\fP tmpCatalyst_ID, \fBacs_longInt\fP tmpSubstrate_ID, \fBacs_double\fP tmpVolume, \fBacs_double\fP tmpK_phospho, \fBacs_int\fP tmpEnergizable)"
.br
.RI "\fIThis constructor is used to create a molecular complex\&. \fP"
.ti -1c
.RI "\fB~species\fP ()"
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetID\fP () const "
.br
.ti -1c
.RI "string \fBgetSequence\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetSequenceLength\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetAmount\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetNOTchargeMols\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetChargeMols\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetConcentration\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetLoadedConcentration\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetAge\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetReborns\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetDiffusionEnh\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetSolubility\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetComplexDegEnh\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetComplexCutPnt\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetEvaluated\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetCatalyst_ID\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetSubstrate_ID\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetK_phospho\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetEnergizable\fP () const "
.br
.ti -1c
.RI "bool \fBgetConcentrationFixed\fP () const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetFirstConcentration\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetLastSpeciesEvaluated\fP () const "
.br
.ti -1c
.RI "vector< \fBacs_longInt\fP > \fBgetSecSubList\fP () const "
.br
.ti -1c
.RI "vector< \fBacs_double\fP > \fBgetSec_k_SubList\fP () const "
.br
.ti -1c
.RI "vector< \fBacs_longInt\fP > \fBgetCatalysisIfCpx\fP () const "
.br
.ti -1c
.RI "\fBacs_int\fP \fBgetSecSubListSize\fP () const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetSecSubListID\fP (\fBacs_int\fP tmpID) const "
.br
.ti -1c
.RI "\fBacs_double\fP \fBgetSec_k_SubListID\fP (\fBacs_int\fP tmpID) const "
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetCatalysisIfCpxID\fP (\fBacs_int\fP tmpID) const "
.br
.ti -1c
.RI "void \fBincrement\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBspecificIncrement\fP (\fBacs_int\fP tmpIncrement, \fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBsetAmount\fP (\fBacs_int\fP tmpAmount, \fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBsetConcentration\fP (\fBacs_double\fP tmpConc, \fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBdecrement\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "bool \fBsetChargeMols\fP (\fBacs_int\fP tmpMolsToCharge)"
.br
.ti -1c
.RI "bool \fBsetSpecificChargeMols\fP (\fBacs_int\fP tmpMolsToCharge)"
.br
.ti -1c
.RI "bool \fBchargeMol\fP ()"
.br
.ti -1c
.RI "bool \fBunchargeMol\fP ()"
.br
.ti -1c
.RI "void \fBsetEvaluated\fP ()"
.br
.ti -1c
.RI "void \fBsetDiffusion\fP (\fBacs_double\fP tmpDiff)"
.br
.ti -1c
.RI "void \fBsetSolubility\fP (\fBacs_int\fP tmpSol)"
.br
.ti -1c
.RI "void \fBsetKphospho\fP (\fBacs_double\fP tmpKphospho)"
.br
.ti -1c
.RI "void \fBsetNewAge\fP (\fBacs_double\fP tmpLastTimeInterval)"
.br
.ti -1c
.RI "void \fBrebornsIncrement\fP ()"
.br
.ti -1c
.RI "void \fBconcToNum\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBnumToConc\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBresetAge\fP ()"
.br
.ti -1c
.RI "void \fBresetReborns\fP ()"
.br
.ti -1c
.RI "void \fBresetToInitConc\fP (\fBacs_double\fP tmpVolume)"
.br
.ti -1c
.RI "void \fBsetLastSpeciesEvaluated\fP (\fBacs_int\fP tmpID)"
.br
.ti -1c
.RI "void \fBinsertSecSub\fP (\fBacs_longInt\fP tmpID, \fBacs_double\fP tmpK, \fBacs_longInt\fP tmpCat)"
.br
.ti -1c
.RI "bool \fBcheckIFtheSecondSubstrateIsAlreadyPresent\fP (\fBacs_longInt\fP tmpSecSubID)"
.br
.ti -1c
.RI "void \fBinsertGillID\fP (\fBacs_longInt\fP tmpID)"
.br
.ti -1c
.RI "\fBacs_longInt\fP \fBgetGillIDpos\fP (\fBacs_longInt\fP tmpID) const "
.br
.ti -1c
.RI "void \fBshowGillEngagement\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
This class contains declarations of the species class\&. 

class species 
.PP
\fBAuthors:\fP
.RS 4
Alessandro Filisetti 
.RE
.PP
\fBVersion:\fP
.RS 4
1\&.1 questa modifica è di prova per subversion
.RE
.PP
Created by Alessandro Filisetti on 19/02/09\&. Copyright 2009 European Centre for Living Technology\&. All rights reserved\&. Test paxelito SVN 
.PP
Definition at line 18 of file species\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "species::species ()"

.PP
< New species constructor (IN AMOUNT) This class containing the declaration of the species\&.
.PP
class species 
.PP
\fBAuthors:\fP
.RS 4
Alessandro Filisetti 
.RE
.PP
\fBVersion:\fP
.RS 4
0\&.1
.RE
.PP
Created by Alessandro Filisetti on 19/02/09\&. Copyright 2009 European Centre for Living Technology\&. All rights reserved\&.Default constructor 
.PP
Definition at line 16 of file species\&.cpp\&.
.SS "species::species (\fBacs_longInt\fPtmpID, stringtmpSequence, \fBacs_longInt\fPtmpAmount, \fBacs_double\fPtmpDiffusionEnh, \fBacs_int\fPtmpSoluble, \fBacs_double\fPtmpComplexDegEnh, \fBacs_int\fPtmpComplexCuttingPoint, \fBacs_int\fPtmpEvalueted, \fBacs_double\fPtmpVolume, \fBacs_double\fPtmpK_phospho, \fBacs_int\fPtmpEnergizable, \fBacs_double\fPtmpInflux_rate, \fBacs_int\fPtmpMaxLOut)"

.PP
New species constructor (IN CONCENTRATION) This constructor is used each time a new species is created (AMOUNT BASED)
.PP
\fBParameters:\fP
.RS 4
\fItmpID\fP species identificator 
.br
\fItmpSequence\fP species sequence (e\&.g\&. ABABAABABA) 
.br
\fItmpAmount\fP species initial amount 
.br
\fItmpDiffusionEnh\fP Diffusion enhancement degree 
.br
\fItmpSoluble\fP 1 if the species is soluble, 0 otherwise 
.br
\fItmpComplexDegEnh\fP complex dissociation kinetic constant 
.br
\fItmpComplexCuttingPoint\fP complex cutting point (catalyst-substrate) 
.br
\fItmpEvalueted\fP This parameter indicates whether the species has been already evalutad (i\&.e\&. all the catalysis of the species are instantiated) 
.br
\fItmpVolume\fP the volume is necessary to convert numbers in concentrations 
.br
\fItmpK_phospho\fP phosphorilation kinetic constant (in case of energy based simulations) 
.br
\fItmpEnergizable\fP this is a flag indicating whether or not the species is energizable 
.RE
.PP

.PP
Definition at line 53 of file species\&.cpp\&.
.SS "species::species (\fBacs_longInt\fPtmpID, stringtmpSequence, \fBacs_double\fPtmpConcentration, \fBacs_double\fPtmpDiffusionEnh, \fBacs_int\fPtmpSoluble, \fBacs_double\fPtmpComplexDegEnh, \fBacs_int\fPtmpComplexCuttingPoint, \fBacs_int\fPtmpEvalueted, \fBacs_double\fPtmpVolume, \fBacs_double\fPtmpK_phospho, \fBacs_int\fPtmpEnergizable, \fBacs_double\fPtmpInflux_rate, \fBacs_int\fPtmpMaxLOut)"

.PP
New species constructor in case of species structure file upload (IN AMOUNT) This constructor is used each time a new species is created (CONCENTRATION BASED)
.PP
\fBParameters:\fP
.RS 4
\fItmpID\fP species identificator 
.br
\fItmpSequence\fP species sequence (e\&.g\&. ABABAABABA) 
.br
\fItmpConcentration\fP species initial concentration 
.br
\fItmpDiffusionEnh\fP Diffusion enhancement degree 
.br
\fItmpSoluble\fP 1 if the species is soluble, 0 otherwise 
.br
\fItmpComplexDegEnh\fP complex dissociation kinetic constant 
.br
\fItmpComplexCuttingPoint\fP complex cutting point (catalyst-substrate) 
.br
\fItmpEvalueted\fP This parameter indicates whether the species has been already evalutad (i\&.e\&. all the catalysis of the species are instantiated) 
.br
\fItmpVolume\fP the volume is necessary to convert concentrations in numbers 
.br
\fItmpK_phospho\fP phosphorilation kinetic constant (in case of energy based simulations) 
.br
\fItmpEnergizable\fP this is a flag indicating whether or not the species is energizable 
.RE
.PP

.PP
Definition at line 99 of file species\&.cpp\&.
.SS "species::species (\fBacs_longInt\fPtmpID, stringtmpSequence, \fBacs_longInt\fPtmpAmount, \fBacs_double\fPtmpDiffusionEnh, \fBacs_int\fPtmpSoluble, \fBacs_double\fPtmpComplexDegEnh, \fBacs_int\fPtmpComplexCuttingPoint, \fBacs_int\fPtmpEvalueted, \fBacs_double\fPtmpAge, \fBacs_int\fPtmpReborns, \fBacs_double\fPtmpVolume, \fBacs_longInt\fPtmpNotUsedCatID, \fBacs_longInt\fPtmpNotUsedSubID, \fBacs_double\fPtmpK_phospho, \fBacs_int\fPtmpEnergizable, \fBacs_double\fPtmpInflux_rate, \fBacs_int\fPtmpMaxLOut)"

.PP
New species constructor in case of species structure file upload (IN CONCENTRATION) This constructor is used when a new species is uploaded from file (TOTAL AMOUNT BASED)
.PP
\fBParameters:\fP
.RS 4
\fItmpID\fP species identificator 
.br

species.cpp
/** class species
 *	\brief This class containing the declaration of the species
 *
 *  @authors Alessandro Filisetti
 *  @version 0.1
 *
 *  Created by Alessandro Filisetti on 19/02/09.
 *  Copyright 2009 European Centre for Living Technology. All rights reserved.
 *
 */

#include "species.h"

/** \brief Default constructor
*/
species::species()
{
	id = 0;
	sequence = "NRG";
	amount = 0;
	chargedMols = 0;
	concentration = 0;
	age = 0;
	reborns = 0;
	diffusionEnh = 1;
	soluble = SOLUBLE;
	complexCuttingPoint = 0;
	complexDegradationEnh = 0;
	evaluated = 1;
	catalyst_ID = 0;
	substrate_ID = 0;
	K_phospho = 0;
	energizable = 0;
	concentrationFixed = false;
	lastSpeciesEvaluated = 0;
	firstConcentration = 0; 

}

/** \brief This constructor is used each time a new species is created (AMOUNT BASED)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param tmpAmount species initial amount
 @param tmpDiffusionEnh Diffusion enhancement degree 
 @param tmpSoluble 1 if the species is soluble, 0 otherwise
 @param tmpComplexDegEnh complex dissociation kinetic constant
 @param tmpComplexCuttingPoint complex cutting point (catalyst-substrate)
 @param tmpEvalueted This parameter indicates whether the species has been already evalutad (i.e. all the catalysis of the species are instantiated)
 @param tmpVolume the volume is necessary to convert numbers in concentrations
 @param tmpK_phospho phosphorilation kinetic constant (in case of energy based simulations)
 @param tmpEnergizable this is a flag indicating whether or not the species is energizable
*/
species::species(acs_longInt tmpID, string tmpSequence, acs_longInt tmpAmount, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpComplexDegEnh, acs_int tmpComplexCuttingPoint, 
				 acs_int tmpEvalueted, acs_double tmpVolume, acs_double tmpK_phospho, acs_int tmpEnergizable, acs_double tmpInflux_rate, acs_int tmpMaxLOut)
{
	id = tmpID;
	sequence = tmpSequence;
	amount = tmpAmount;
	chargedMols = 0;
	numToConc(tmpVolume);
	age = 0;
	reborns = 0;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	complexDegradationEnh = tmpComplexDegEnh;
	complexCuttingPoint = tmpComplexCuttingPoint;
	evaluated = tmpEvalueted;
	catalyst_ID = 0;
	substrate_ID = 0;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	
	// Set concentrationFixed propriety
	concentrationFixed = false;
	if((tmpInflux_rate == 0) && (tmpMaxLOut > 0))
	{
		if(sequence.length() <= tmpMaxLOut) concentrationFixed = true;
	}

    firstConcentration = 0; // 0 because the species is created during on run
    lastSpeciesEvaluated = 0;

}

/** \brief This constructor is used each time a new species is created (CONCENTRATION BASED)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param tmpConcentration species initial concentration
 @param tmpDiffusionEnh Diffusion enhancement degree 
 @param tmpSoluble 1 if the species is soluble, 0 otherwise
 @param tmpComplexDegEnh complex dissociation kinetic constant
 @param tmpComplexCuttingPoint complex cutting point (catalyst-substrate)
 @param tmpEvalueted This parameter indicates whether the species has been already evalutad (i.e. all the catalysis of the species are instantiated)
 @param tmpVolume the volume is necessary to convert concentrations in numbers
 @param tmpK_phospho phosphorilation kinetic constant (in case of energy based simulations)
 @param tmpEnergizable this is a flag indicating whether or not the species is energizable
 */
species::species(acs_longInt tmpID, string tmpSequence, acs_double tmpConcentration, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpComplexDegEnh, acs_int tmpComplexCuttingPoint, 
				 acs_int tmpEvalueted, acs_double tmpVolume, acs_double tmpK_phospho, acs_int tmpEnergizable, acs_double tmpInflux_rate, acs_int tmpMaxLOut)
{
	id = tmpID;
	sequence = tmpSequence;
	concentration = tmpConcentration;
	concToNum(tmpVolume);
	numToConc(tmpVolume);
	chargedMols = 0;
	age = 0;
	reborns = 0;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	complexDegradationEnh = tmpComplexDegEnh;
	complexCuttingPoint = tmpComplexCuttingPoint;
	evaluated = tmpEvalueted;
	catalyst_ID = 0;
	substrate_ID = 0;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	
	// Set concentrationFixed propriety
	concentrationFixed = false;
	if((tmpInflux_rate == 0) && (tmpMaxLOut > 0))
	{
		if(sequence.length() <= tmpMaxLOut) concentrationFixed = true;
	}

    firstConcentration = 0; // 0 because the species is created during on run
    lastSpeciesEvaluated = 0;
}

/** \brief This constructor is used when a new species is uploaded from file (TOTAL AMOUNT BASED)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param tmpAmount species initial amount of molecules
 @param tmpReactions_constant ????
 */
species::species(acs_longInt tmpID, string tmpSequence, acs_longInt tmpAmount, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpComplexDegEnh, acs_int tmpComplexCuttingPoint, 
				 acs_int tmpEvalueted, acs_double tmpAge, acs_int tmpReborns, acs_double tmpVolume, 
                 acs_longInt tmpNotUsedCatID, acs_longInt tmpNotUsedSubID, acs_double tmpK_phospho,
                 acs_int tmpEnergizable, acs_double tmpInflux_rate, acs_int tmpMaxLOut)
{

	id = tmpID;
	sequence = tmpSequence;
	amount = tmpAmount;
	chargedMols = 0;
    numToConc(tmpVolume); // The concetration is set within this function
	age = tmpAge;
	reborns = tmpReborns;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	complexDegradationEnh = tmpComplexDegEnh;
	complexCuttingPoint = tmpComplexCuttingPoint;
	evaluated = tmpEvalueted;
	catalyst_ID = tmpNotUsedCatID;
	substrate_ID = tmpNotUsedSubID;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	
	// Set concentrationFixed propriety
	concentrationFixed = false;
	if((tmpInflux_rate == 0) && (tmpMaxLOut > 0))
	{
		if(sequence.length() <= tmpMaxLOut) concentrationFixed = true;
	}

    firstConcentration = concentration; // If the species is loaded from file that's the very initial concentration
    lastSpeciesEvaluated = 0;
	
}

/** \brief This constructor is used when a new species is uploaded from file (CONCENTRATION BASED)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param tmpAmount species initial amount of molecules
 @param tmpReactions_constant ????
 */
species::species(acs_longInt tmpID, string tmpSequence, acs_double tmpConcentration, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpComplexDegEnh, acs_int tmpComplexCuttingPoint, 
				 acs_int tmpEvalueted, acs_double tmpAge, acs_int tmpReborns, acs_double tmpVolume, 
                 acs_longInt tmpNotUsedCatID, acs_longInt tmpNotUsedSubID, acs_double tmpK_phospho,
                 acs_double tmpKLoadConc, acs_int tmpEnergizable, acs_double tmpInflux_rate, acs_int tmpMaxLOut)

{

	id = tmpID;
	sequence = tmpSequence;
	concentration = tmpConcentration;
	concToNum(tmpVolume);
	numToConc(tmpVolume);
	chargedMols = round(tmpKLoadConc*tmpVolume*AVO);
	age = tmpAge;
	reborns = tmpReborns;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	complexDegradationEnh = tmpComplexDegEnh;
	complexCuttingPoint = tmpComplexCuttingPoint;
	evaluated = tmpEvalueted;
	catalyst_ID = tmpNotUsedCatID;
	substrate_ID = tmpNotUsedSubID;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	
	// Set concentrationFixed propriety
	concentrationFixed = false;
	if((tmpInflux_rate == 0) && (tmpMaxLOut > 0))
	{
		if(sequence.length() <= tmpMaxLOut) concentrationFixed = true;
	}

    firstConcentration = tmpConcentration; // If the species is loaded from file that's the very initial concentration
    lastSpeciesEvaluated = 0;
}

/** \brief This constructor is used when a species is randomly created (!!! NOT USED NOW)
 @version 0.1 (8 parameters)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param tmpAmount species initial amount of molecules
 @param acs_double tmpDiffusionEnh Diffusione enhancement parameter
 @param acs_double tmpPrecipitationEnh Precipitation Enhancement parameters
 @param acs_double tmpComplexProb Probability to be a complex
 @param acs_double tmpMaxComplexDegKinetic max complex degradation constant
 @param MTRand& tmp_RandomGenerator random generator
 */
species::species(acs_longInt tmpID, string tmpSequence, acs_longInt tmpAmount, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpComplexProb, acs_double tmpMaxComplexDegKinetic, 
				 MTRand& tmp_RandomGenerator, acs_double tmpVolume, acs_double tmpK_phospho, acs_int tmpEnergizable)
{
	cout << "four" << endl;

	id = tmpID;
	sequence = tmpSequence;
	amount = tmpAmount;
	numToConc(tmpVolume);
	chargedMols = 0;
	age = 0;
	reborns = 0;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	// RANDOM SELECTION BETWEEN SPECIES AND COMPLEX ACCORDING TO THE COMPLEX CREATION PROBABILITY
	acs_double dice = tmp_RandomGenerator();
	if(dice <= tmpComplexProb) //COMPLEX
	{
		complexDegradationEnh = getDoubleRandom(0.0, tmpMaxComplexDegKinetic, tmp_RandomGenerator);
		complexCuttingPoint = getIntRandom(2, sequence.length(), tmp_RandomGenerator) - 1;
		
	}else{ //SPECIES
		complexDegradationEnh = 0;
		complexCuttingPoint = 0;
	}
	evaluated = 0;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	concentrationFixed = false;
    firstConcentration = 0; // 0 because the species is created during on run
    lastSpeciesEvaluated = 0;
}

/** \brief This constructor is used to create a molecular complex
 @version 0.1 (10 paramters)
 @param tmpID species identificator
 @param tmpSequence species sequence (e.g. ABABAABABA)
 @param acs_double tmpDiffusionEnh Diffusione enhancement parameter
 @param acs_double tmpPrecipitationEnh Precipitation Enhancement parameters
 @param acs_double tmpComplexProb Probability to be a complex
 @param acs_double tmpMaxComplexDegKinetic max complex degradation constant
 @param MTRand& tmp_RandomGenerator random generator
 @param acs_int tmpCatalyst_ID Catalyst ID
 @param acs_int tmpSubstrate_ID substrate ID
 */
species::species(acs_longInt tmpID, string tmpSequence, acs_double tmpDiffusionEnh, 
				 acs_int tmpSoluble, acs_double tmpMaxComplexDegKinetic, 
				 acs_int tmpCuttingPoint, MTRand& tmp_RandomGenerator, acs_longInt tmpCatalyst_ID, 
				 acs_longInt tmpSubstrate_ID, acs_double tmpVolume, acs_double tmpK_phospho, acs_int tmpEnergizable, acs_int tmpAmount)
{
	id = tmpID;
	sequence = tmpSequence;
	amount = tmpAmount;
	chargedMols = 0;
	numToConc(tmpVolume);
	age = 0;
	reborns = 0;
	diffusionEnh =tmpDiffusionEnh;
	soluble = tmpSoluble;
	complexDegradationEnh = getDoubleRandom(tmpMaxComplexDegKinetic, tmpMaxComplexDegKinetic, tmp_RandomGenerator);
	complexCuttingPoint = tmpCuttingPoint;
	evaluated = 0;
	catalyst_ID = tmpCatalyst_ID;
	substrate_ID = tmpSubstrate_ID;
	K_phospho = tmpK_phospho;
	energizable = tmpEnergizable;
	concentrationFixed = false;
    firstConcentration = 0; // 0 because the species is created during on run
    lastSpeciesEvaluated = 0;
}

/** Function to increment the total number of molecules belonging to this species
 */
void species::increment(acs_double tmpVolume)
{
    if(!concentrationFixed)
   {
       amount++;
       numToConc(tmpVolume);
   }
}

/** Function to decrement the total number of molecules belonging to this species
 */
void species::decrement(acs_double tmpVolume)
{
   if(!concentrationFixed)
   {
       amount--;
       numToConc(tmpVolume);
   }
}

/** \Funtion to charge a specific number of molecules
 */
bool species::setSpecificChargeMols(acs_int tmpMolsToCharge)
{
	bool chargeOK = false;
	chargedMols += tmpMolsToCharge;
	if (chargedMols > amount) {
		chargedMols = amount;
		chargeOK = true;
	}
	return chargeOK;
}

/** \Funtion to charge a specific number of molecules
 */
bool species::setChargeMols(acs_int tmpMolsToCharge)
{
	bool chargeOK = false;
	chargedMols = tmpMolsToCharge;
	if (chargedMols > amount) {
		chargedMols = amount;
		chargeOK = true;
	}
	return chargeOK;
}

/** \Funtion to charge molecules
*/
bool species::chargeMol()
{
	bool chargeOK = false;
	if (chargedMols < amount) {
		chargedMols++;
		chargeOK = true;
	}else{
		cout << "+: " << chargedMols << " tot: " << amount << endl;
	}

	return chargeOK;
}

/** \Funtion to uncharge molecules
 */
bool species::unchargeMol()
{
	bool unchargeOK = false;
	if (chargedMols > 0) {
		chargedMols--;
		unchargeOK = true;
	}
	return unchargeOK;
}

/** \Function return the concentration of the loaded molecules 
*/
acs_double species::getLoadedConcentration(acs_double tmpVolume)
{
	acs_double tmpConc = chargedMols / (AVO * tmpVolume);
	//cout << "Amount: " << amount << " - Conc:" << concentration << " - Loaded M:" << chargedMols << " -  Conc C:" << tmpConc << endl;
	return tmpConc;
}
/**
 Function to insert the second substrate, final condensation rate and the catalysis ID to the species (complexes list)
 @author Alessandro Filisetti
 @date 2013-12-10
 @param acs_longInt tmpID second substrate ID
 @param acs_double tmpK final condensation ID
 @param acs_longInt tmpCat catalysis ID
 */
void species::insertSecSub(acs_longInt tmpID, acs_double tmpK, acs_longInt tmpCat)
{
    if(secondSubstrates.size() > 0)
    {
    	bool tmpFound = false;
    	for(acs_int z = 0; z < (acs_int)secondSubstrates.size(); z++)
    	{
    		if(secondSubstrates.at(z) == tmpID){tmpFound = true; break;}
    	}
    	if(!tmpFound){secondSubstrates.push_back(tmpID); kCond.push_back(tmpK); catalysisIfCpx.push_back(tmpCat);}
    }else{
    	secondSubstrates.push_back(tmpID);
    	kCond.push_back(tmpK);
    	catalysisIfCpx.push_back(tmpCat);
    }
}

/**
 * Function to show the species gillespie engagement
 */
void species::showGillEngagement()
{
	cout << "|- SPECIES " << id << " is engaged in Gillespie reactions: ";
	for(acs_int i = 0; i < gillespieEngagement.size(); i++)
	{
		cout << gillespieEngagement.at(i) << "\t";
	}
	cout << endl;
}

/**
  Function to control if the second substrare is already present
 */
bool species::checkIFtheSecondSubstrateIsAlreadyPresent(acs_longInt tmpSecSubID)
{
	bool alreadyPresent = false;
	if (secondSubstrates.size() > 0)
	{
		for(vector<acs_longInt>::iterator tmpSecSubIter = secondSubstrates.begin(); tmpSecSubIter != secondSubstrates.end(); tmpSecSubIter++)
		{
			if (tmpSecSubID == *tmpSecSubIter)
				alreadyPresent = true;
		}
	}
	return alreadyPresent;
}


/**
	Function to insert event in list
*/
void species::insertEvent(acs_longInt IDEvent, bool IncOrDec) {

	vector<acs_longInt> *events;
	if (IncOrDec == INC) events = &eventsInc;
	else events = &eventsDec;
	(*events).push_back(IDEvent);
}


/**
	Function to print events in both lists
*/
void species::printEventsList() {

	cout<<"Inc. Events:  ";
	if (eventsInc.size() == 0)
		cout<<"Empty";
	else
		for (acs_longInt i = 0; i < eventsInc.size(); i++)
			cout<<eventsInc[i]<<"  ";

	cout<<endl;

	cout<<"Dec. Events:  ";
	if (eventsDec.size() == 0)
		cout<<"Empty";
	else
		for (acs_longInt i = 0; i < eventsDec.size(); i++)
			cout<<eventsDec[i]<<"  ";

	cout<<endl;

}


